File: typescript-playwright/src/pages/PhotoPage.ts
import type { Page, Locator } from "@playwright/test";
import { BasePage } from "./BasePage.ts";
import { step } from "../utils/decorators.ts";

export class PhotoPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  get photo(): Locator {
    return this.page.locator('[data-testid="photo-img-large"]');
  }
  get likeButton(): Locator {
    return this.page.locator('[data-testid="photo-like"]');
  }
  get viewsCount(): Locator {
    return this.page.locator('[data-testid="photo-views"]');
  }
  get carouselNext(): Locator {
    return this.page.locator('[data-testid="carousel-next"]');
  }
  get carouselPrev(): Locator {
    return this.page.locator('[data-testid="carousel-prev"]');
  }
  get commentsEmpty(): Locator {
    return this.page.locator('[data-testid="comments-empty"]');
  }
  get photoMeta(): Locator {
    return this.page.locator('[data-testid="photo-meta"]');
  }
  get photoTitle(): Locator {
    return this.page.locator('[data-testid="photo-title"]');
  }
  get photoPlace(): Locator {
    return this.page.locator('[data-testid="photo-place"]');
  }
  get photoDate(): Locator {
    return this.page.locator('[data-testid="photo-date"]');
  }
  get lightbox(): Locator {
    return this.page.locator('[data-testid="lightbox"]');
  }
  get closeLightboxButton(): Locator {
    return this.page.locator('[data-testid="close-btn"]');
  }

  private commentItem(commentId: number): Locator {
    return this.page.locator(`[data-testid="comment-item-${commentId}"]`);
  }

  commentByAuthor(author: string): Locator {
    return this.page
      .locator('[data-testid^="comment-author-"]')
      .filter({ hasText: author });
  }

  commentByText(text: string): Locator {
    return this.page
      .locator('[data-testid^="comment-text-"]')
      .filter({ hasText: text });
  }

  photoByAlt(altText: string): Locator {
    return this.page.getByAltText(altText);
  }

  @step("Открыть страницу фото")
  async open(photoId: number): Promise<void> {
    await this.page.goto(`/photo/${photoId}`);
  }

  @step("Перейти к следующему фото")
  async goNext(): Promise<void> {
    await this.carouselNext.click();
  }

  @step("Перейти к предыдущему фото")
  async goPrev(): Promise<void> {
    await this.carouselPrev.click();
  }

  @step("Дождаться отображения комментария")
  async waitForCommentVisible(commentId: number): Promise<void> {
    await this.commentItem(commentId).waitFor({ state: "visible" });
  }
}

File: python/src/pages/photo_page.py
import allure
from playwright.async_api import Locator, Page

from src.pages.base_page import BasePage
from src.utils.constants import STATE_VISIBLE


class PhotoPage(BasePage):
    URL_PATH = "/photo"

    def __init__(self, page: Page):
        super().__init__(page)

    @property
    def photo(self) -> Locator:
        return self.page.get_by_test_id("photo-img-large")

    @property
    def like_btn(self) -> Locator:
        return self.page.get_by_test_id("photo-like-btn")

    @property
    def views_count(self) -> Locator:
        return self.page.get_by_test_id("photo-views-count")

    @property
    def carousel_next_btn(self) -> Locator:
        return self.page.get_by_test_id("carousel-next-btn")

    @property
    def carousel_prev_btn(self) -> Locator:
        return self.page.get_by_test_id("carousel-prev-btn")

    @property
    def comments_empty_state(self) -> Locator:
        return self.page.get_by_test_id("comments-empty-state")

    @property
    def photo_meta(self) -> Locator:
        return self.page.get_by_test_id("photo-meta")

    @property
    def carousel_counter(self) -> Locator:
        return self.page.get_by_test_id("carousel-counter")

    @property
    def carousel_counter_current(self) -> Locator:
        return self.page.get_by_test_id("carousel-counter-current")

    @property
    def photo_title(self) -> Locator:
        return self.page.get_by_test_id("photo-title")

    @property
    def photo_place(self) -> Locator:
        return self.page.get_by_test_id("photo-place")

    @property
    def photo_date(self) -> Locator:
        return self.page.get_by_test_id("photo-date")

    @property
    def lightbox(self) -> Locator:
        return self.page.get_by_test_id("lightbox")

    @property
    def close_lightbox_btn(self) -> Locator:
        return self.page.get_by_test_id("close-lightbox-btn")

    def comment_item(self, comment_id: int) -> Locator:
        return self.page.get_by_test_id(f"comment-item-{comment_id}")

    def comment_by_author(self, author: str) -> Locator:
        return self.page.locator('[data-testid^="comment-author-"]').filter(has_text=author)

    def comment_by_text(self, text: str) -> Locator:
        return self.page.locator('[data-testid^="comment-text-"]').filter(has_text=text)

    def photo_by_alt(self, alt: str) -> Locator:
        return self.page.get_by_alt_text(alt)

    @allure.step("Открыть страницу фото с ID: {photo_id}")
    async def open(self, photo_id: int) -> None:
        await self.page.goto(f"{self.URL_PATH}/{photo_id}")

    @allure.step("Перейти к следующему фото")
    async def go_next(self) -> None:
        await self.carousel_next_btn.click()

    @allure.step("Перейти к предыдущему фото")
    async def go_prev(self) -> None:
        await self.carousel_prev_btn.click()

    @allure.step("Дождаться отображения комментария: {comment_id}")
    async def wait_for_comment_is_visible(self, comment_id: int) -> None:
        await self.comment_item(comment_id).wait_for(state=STATE_VISIBLE)
